### ISSUE TITLE
Eager context seems contradicting the distributed TensorFlow concept

### ISSUE THREAD

**Issue by pw2393** [tensorflow/tensorflow#34242]
**System information**
- OS Platform and Distribution: Ubuntu 18.04
- TensorFlow installed from (source or binary): binary whl
- TensorFlow version (use command below): `tensorflow-gpu==2.0.0`
- Python version: 3.6.8
- CUDA/cuDNN version: 10.0 / 7.6.4
- GPU model and memory: GeForce GTX 1080 Ti

**Describe the current behavior**
```python
import argparse

import tensorflow as tf
from tensorflow.core.protobuf.tensorflow_server_pb2 import ServerDef
from tensorflow.python.eager import context
from tensorflow.python.training.server_lib import ClusterSpec


CLUSTER_SPEC = {
    "worker": [
        "localhost:14286",
        "localhost:14287"
    ]
}


def test():
    c = context.context()
    c.set_server_def(
        ServerDef(
            cluster=ClusterSpec(CLUSTER_SPEC).as_cluster_def(),
            job_name=FLAGS.job_name,
            task_index=FLAGS.task_index,
            protocol='grpc'
        )
    )
    print(tf.constant(3))
    print('lol')


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--job_name",
        type=str,
    )
    parser.add_argument(
        "--task_index",
        type=int,
    )
    FLAGS, _ = parser.parse_known_args()
    test()
```

We refer the above code as `a.py`.
STEP 1: we launch it by
```bash
python a.py --job_name=worker --task_index=0
```
with the log
```
...
2019-11-13 13:53:38.096726: I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:258] Initialize GrpcChannelCache for job worker -> {0 -> localhost:14286, 1 -> localhost:14287}
2019-11-13 13:53:38.099209: I tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc:365] Started server with target: grpc://localhost:14286
```
and it is waiting for another server as described in the `CLUSTER_SPEC`. Note that no `print` above has been executed yet.

STEP 2: we launch it as another process by
```bash
python a.py --job_name=worker --task_index=1
```
with the log
```
...
2019-11-13 14:00:01.431825: I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:258] Initialize GrpcChannelCache for job worker -> {0 -> localhost:14286, 1 -> localhost:14287}
2019-11-13 14:00:01.433641: I tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc:365] Started server with target: grpc://localhost:14287
2019-11-13 14:00:01.471394: I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:258] Initialize GrpcChannelCache for job worker -> {0 -> localhost:14286, 1 -> localhost:14287}
tf.Tensor(3, shape=(), dtype=int32)
lol
2019-11-13 14:00:01.585211: E tensorflow/core/distributed_runtime/rpc/eager/grpc_eager_client.cc:72] Remote EagerContext with id 2379290034739913425 does not seem to exist.
```
and the first processor with the `task_index=0` hangs forever.

**Describe the expected behavior**

In TensorFlow graph mode, we know when multiple sessions talk to the same cluster (which is defined by `ClusterSpec`), everything works pretty smooth and natural. 

Should eager context behave similarly?  That is, both the above processes are expected to run naturally: print the contant and string and exited.

In other words, when multiple contexts talking to the same cluster (which is defined by `ClusterSpec`), should they know how to coordinate with each other, just as the TensorFlow core is architectured and designed before?

(Or is there any other API that could help resolve this issue?)

**Code to reproduce the issue**
See above

**Comment by mohantym** [comment]
Was able to replicate the issue with TF v2.5 ,after removing "`print(tf.constant(3)`)" for sake of simplicity, was able to run both session without hanging first session,please find the [gist ](https://colab.research.google.com/gist/mohantym/9c1a3d1f4f078bf0efbd4e6d8a8db27a/34242.ipynb#scrollTo=RVf58TTSs_UQ)here ..Thanks!
